src/input.rs

```diff
@@ -197,6 +202,31 @@ where
         (self.mouse_rel[0], self.mouse_rel[1])
     }
 
+    pub fn set_text_input_enabled(&mut self, enabled: bool) {
+        self.text_input_enabled = enabled;
+    }
+
+    pub fn text_input_update(&mut self, s: &mut String) -> bool {
+        let mut found_return = false;
+        for (keycode, keymod) in std::mem::take(&mut self.text_input_events) {
+            match keycode {
+                Keycode::Backspace => { s.pop(); }
+                Keycode::Escape => { s.clear(); found_return = true; }
+                Keycode::Return => found_return = true,
+                c => if let Some(mut c) = (c as i32).try_into().ok().and_then(char::from_u32) {
+                    //TODO(gu): Handle non-ascii better.
+                    if c.is_ascii_alphanumeric() || c.is_ascii_punctuation() || c == ' ' {
```

Comment:
is there an `is ascii`?

Reply:
Yes, but that also allows the control characters.